runtime/lua/lsp/callbacks.lua

```diff
@@ -0,0 +1,566 @@
+end
+
+-- CallbackObject section
+CallbackObject.__call = function(self, success, data, default_only, filetype)
+  if self.name ~= 'nvim/error_callback' and not success then
+    call_callbacks_for_method('nvim/error_callback', false, data)
+  end
+
+  if util.table.is_empty(self.default) then
+      log.trace('Request: "', self.method, '" had no registered callbacks')
+    return nil
+  end
+
+  local callback_list = self:get_list_of_callbacks(default_only, filetype)
+  local results = { }
+
+  for _, cb in ipairs(callback_list) do
+    local current_result = cb(self, data)
+
+    if current_result ~= nil then
+      table.insert(results, current_result)
+    end
+  end
+
+  return unpack(results)
+end
+
+CallbackObject.new = function(method, default_callback, options)
+  options = options or {}
+
+  local object = setmetatable({
+    method = method,
+    options = setmetatable(options, {
+      __index = function(_, key)
+        error(string.format('Cannot get option (%s) for "%s"', key, method))
+      end,
+
+      __newindex = function(_, key, value)
+        error(string.format('Cannot set option (%s) for "%s" to value [%s]', key, method, value))
+      end,
+    }),
+
+    default = {},
+    generic = {},
+    filetype = {},
+  }, CallbackObject)
+
+  if default_callback ~= nil then
+    if type(default_callback) == 'string' then
+      default_callback = function(self, data)
+        return vim.api.nvim_call_function(default_callback, {
+          { method = self.method, options = unpack(options) },
+          data
+        })
+      end
+    end
+
+    object:set_default_callback(default_callback)
+  end
+
+  return object
+end
+
+CallbackObject.set_default_callback = function(self, default_callback)
+  self.default = { default_callback }
+end
+
+CallbackObject.add_callback = function(self, new_callback)
+  table.insert(self.generic, new_callback)
+end
+
+CallbackObject.add_default_callback = function(self, new_callback)
+  table.insert(self.default, new_callback)
+end
+
+CallbackObject.add_filetype_callback = function(self, new_callback, filetype)
+  if self.filetype[filetype] == nil then
+    self.filetype[filetype] = {}
+  end
+
+  table.insert(self.filetype[filetype], new_callback)
+end
+
+CallbackObject.get_list_of_callbacks = function(self, default_only, filetype)
+  local callback_list = {}
+
+  for _, value in ipairs(self.default) do
+    table.insert(callback_list, value)
+  end
+
+  if default_only then
+    return callback_list
+  end
+
+  for _, value in ipairs(self.generic) do
+    table.insert(callback_list, value)
+  end
+
+  if filetype ~= nil then
+    if self.filetype[filetype] == nil then
+      self.filetype[filetype] = {}
+    end
+
+    for _, value in ipairs(self.filetype[filetype]) do
+      table.insert(callback_list, value)
+    end
+  end
+
+  return callback_list
+end
+
+-- Operation function for CallbackMapping and CallbackObject section
+local get_callback_object_by_method = function(method, create_new)
```

Comment:
There are many functions that have side effects on global objects like this one. Is this normal in Lua?

Reply:
You mean that it's mutating `CallbackMapping`? That's not really a global object, but local to that file. Doing it this way is indeed pretty typical for lua code. Note that `CallbackMapping` is not exported, so all usage of it is restricted to that file.

Reply:
Oh, I see. I've misunderstood Lua scope :-)